include/small/lsregion.h

```diff
@@ -68,6 +69,10 @@ struct lslab {
 	 * structure.
 	 */
 	size_t slab_used;
+	/**
+	 * The first id that was used to alloc data from the slab.
+	 */
+	int64_t min_id;
```

Comment:
Are you sure you need this member? Wouldn't this also work?
```C
struct svp {
    int64_t max_id;
    size_t offset;
};

/**
 * out_svp - out-parameter, telling max_id of last lslab saved into iovecs.
 *     and offset means "skip this many first bytes, they were already sent", and update
 *         the offset to whatever was the new last slab's size again.
 * max_count - max number of vectors to fill.
 * Returns number of vectors actually filled.
 */
int to_iovec(lsregion *region, svp *out_svp, iovec *vecs, int max_count);
```

So you just save how many bytes need to be skipped on next to-iovec from the first lslab. I am trying to lead us to the logic from this place: https://github.com/Gerold103/serverbox/blob/master/src/mg/net/Buffer.cpp#L333:
```C++
	uint32_t
	BuffersToIOVecsForWrite(
		const Buffer* aHead,
		uint32_t aByteOffset,
		mg::box::IOVec* aVectors,
		uint32_t aVectorCount)
	{
		mg::box::IOVec* buf = aVectors;
		mg::box::IOVec* bufEnd = aVectors + aVectorCount;
		for (; aHead != nullptr && buf < bufEnd; aHead = aHead->myNext.GetPointer())
		{
			if (aHead->myPos == 0)
				continue;
			MG_DEV_ASSERT(aByteOffset < aHead->myPos);
			buf->myData = aHead->myWData + aByteOffset;
			buf->mySize = aHead->myPos - aByteOffset;
			++buf;
			aByteOffset = 0;
		}
		return (uint32_t)(buf - aVectors);
	}
```
`aHead` is like a list of lslabs. And `aByteOffset` is how many bytes need to be skipped in the beginning because they were already sent on a previous write.

Reply:
> Are you sure you need this member? Wouldn't this also work?

I'm afraid I do.

First of all, the user isn't obliged to call `lsregion_gc` after `lsregion_to_iovec`, so the next `lsregion_to_iovec` might have to flush the data starting from some intermediate slab (not the first one in the list). If this is the case, I don't see any alternatives to `min_id` and `max_id` pairs to determine the right slab.

Ok, you may say that we'll force the user to call `lsregion_gc` every time, but still using just `max_id` + `pos` won't be enough: imagine you flush a single slab with `max_id = 3` up to `pos = 10`. You'll then have `svp.max_id = 3` and `svp.pos = 10`. If a new allocation with `id = 4` happens between `lsregion_to_iovec` and `lsregion_gc` (for example, the caller fiber yields inside `coio_writev` and some other fiber allocates more data, like this will happen with relay rewriting to multiple fibers), you have two cases:
 * the allocation is small enough to fit in the same slab. `slab.max_id = 4` and slab.pos is propagated as needed. In this case our `lsregion_gc` with `max_id = 3` will not free the slab (it has some extra data, obviously). The next `lsregion_to_iovec` is correct. It sees our slab with `max_id = 4 >= svp.max_id`  and skips the already flushed 10 bytes.
 * the allocation doesn't fit in our slab, so a new one is created. Our slab remains with `max_id = 3` and the new slab has `max_id =4` and some corresponding `pos`. Now `lsregion_gc` with `max_id = 3` deletes our slab, and the next call to `lsregion_to_iovec` becomes incorrect: it has `svp.max_id = 3` which is less than `new_slab.max_id`, so we assume we need to skip the first `svp.pos` bytes of a slab that we have never flushed. We'll essentially skip some not yet flushed data.

There's another alternative that I've thought about: maybe I could save a pointer to the last seen slab. But again this is not the answer, because of `lsregion::cached`. The same slab may be reused after `gc` leading to the same problems.

I couldn't come up with any other ideas than `max_id,min_id` pairs. Do you have any ideas?

Reply:
Oh, okay, you are right. Won't work here. In my other place it works because on "gc" I take offset and return new offset, and always maintain it like a cursor. We probably shouldn't do this here.

Having `min_id` check on each `lslab_use()` still bothers me. If we only want to remember the last used slab in the savepoint, we could perhaps introduce slab ID? Have a monotonic counter, increment and assign it on slab creation/reuse, and store in the svp "last used slab id". So it would be 2 ids in svp: last used alloc ID and last used slab ID. This way we can detect if the offset is valid or not, and `lslab_use()` wouldn't need to have an `if`.

How is that idea?

Reply:
Done:

<details><summary>diff</summary>

```diff
diff --git a/include/small/lsregion.h b/include/small/lsregion.h
index bdc7398..12b9370 100644
--- a/include/small/lsregion.h
+++ b/include/small/lsregion.h
@@ -70,9 +70,9 @@ struct lslab {
 	 */
 	size_t slab_used;
 	/**
-	 * The first id that was used to alloc data from the slab.
+	 * An id given to this slab at creation.
 	 */
-	int64_t min_id;
+	int64_t slab_id;
 	/**
 	 * Maximal id that was used to alloc data from the slab.
 	 */
@@ -108,6 +108,11 @@ struct lsregion {
 	/** Slabs arena - source for memory slabs. */
 	struct slab_arena *arena;
 	struct lslab *cached;
+	/**
+	 * A monotonically growing id of used slabs. Needed for savepoint
+	 * tracking.
+	 */
+	int64_t slab_id;
 };
 
 /** Aligned size of the struct lslab. */
@@ -119,13 +124,13 @@ lslab_sizeof()
 
 /** Initialize the lslab object. */
 static inline void
-lslab_create(struct lslab *slab, size_t size)
+lslab_create(struct lslab *slab, size_t size, int64_t slab_id)
 {
 	rlist_create(&slab->next_in_list);
 	slab->slab_size = size;
 	slab->slab_used = lslab_sizeof();
+	slab->slab_id = slab_id;
 	slab->max_id = LSLAB_NOT_USED_ID;
-	slab->min_id = LSLAB_NOT_USED_ID;
 }
 
 /**
@@ -149,8 +154,6 @@ lslab_use(struct lslab *slab, size_t size, int64_t id)
 	assert(size <= lslab_unused(slab));
 	assert(slab->max_id <= id);
 	slab->slab_used += size;
-	if (slab->min_id == LSLAB_NOT_USED_ID)
-		slab->min_id = id;
 	slab->max_id = id;
 }
 
@@ -178,11 +181,11 @@ lslab_end(struct lslab *slab)
  */
 struct lsregion_svp {
 	/**
-	 * The max allocation id that was seen during a flush. Using it we can
+	 * The max slab id that was seen during a flush. Using it we can
 	 * find the slab we flushed previously and check if it has any extra
 	 * data to flush.
 	 */
-	int64_t max_id;
+	int64_t slab_id;
 	/** A position in the lslab containing an allocation with \a id. */
 	size_t pos;
 };
@@ -190,7 +193,7 @@ struct lsregion_svp {
 static inline void
 lsregion_svp_create(struct lsregion_svp *svp)
 {
-	svp->max_id = LSLAB_NOT_USED_ID;
+	svp->slab_id = LSLAB_NOT_USED_ID;
 	svp->pos = 0;
 }
 
@@ -207,6 +210,7 @@ lsregion_create(struct lsregion *lsregion, struct slab_arena *arena)
 	slab_list_create(&lsregion->slabs);
 	lsregion->arena = arena;
 	lsregion->cached = NULL;
+	lsregion->slab_id = 0;
 }
 
 /** @sa lsregion_aligned_reserve(). */
@@ -346,7 +350,8 @@ lsregion_gc(struct lsregion *lsregion, int64_t min_id)
 			lsregion->slabs.stats.total -= slab->slab_size;
 			slab_unmap(lsregion->arena, slab);
 		} else {
-			lslab_create(slab, slab->slab_size);
+			lslab_create(slab, slab->slab_size,
+				     ++lsregion->slab_id);
 			lsregion->cached = slab;
 		}
 	}
@@ -358,8 +363,9 @@ lsregion_gc(struct lsregion *lsregion, int64_t min_id)
  * @param[in,out] iovcnt Max iovec count.
  *			 Returns how many iovecs were actually used.
  * @param[in,out] svp A savepoint pointing at the end of flushed data.
+ * @return max allocation id that was flushed.
  */
-void
+int64_t
 lsregion_to_iovec(const struct lsregion *lsregion, struct iovec *iov,
 		  int *iovcnt, struct lsregion_svp *svp);
 
diff --git a/small/lsregion.c b/small/lsregion.c
index 348fba0..4c8b594 100644
--- a/small/lsregion.c
+++ b/small/lsregion.c
@@ -67,7 +67,7 @@ lsregion_aligned_reserve_slow(struct lsregion *lsregion, size_t size,
 			quota_release(quota, slab_size);
 			return NULL;
 		}
-		lslab_create(slab, slab_size);
+		lslab_create(slab, slab_size, ++lsregion->slab_id);
 		rlist_add_tail_entry(&lsregion->slabs.slabs, slab,
 				     next_in_list);
 		lsregion->slabs.stats.total += slab_size;
@@ -81,7 +81,7 @@ lsregion_aligned_reserve_slow(struct lsregion *lsregion, size_t size,
 		slab = (struct lslab *) slab_map(arena);
 		if (slab == NULL)
 			return NULL;
-		lslab_create(slab, slab_size);
+		lslab_create(slab, slab_size, ++lsregion->slab_id);
 		rlist_add_tail_entry(&lsregion->slabs.slabs, slab,
 				     next_in_list);
 		lsregion->slabs.stats.total += slab_size;
@@ -92,12 +92,13 @@ lsregion_aligned_reserve_slow(struct lsregion *lsregion, size_t size,
 	return pos;
 }
 
-void
+int64_t
 lsregion_to_iovec(const struct lsregion *lsregion, struct iovec *iov,
 		  int *iovcnt, struct lsregion_svp *svp)
 {
 	struct lslab *lslab;
-	int64_t prev_id = svp->max_id;
+	int64_t prev_id = svp->slab_id;
+	int64_t max_alloc_id = 0;
 	size_t prev_pos = svp->pos;
 	int max_iovcnt = *iovcnt;
 	int cnt = 0;
@@ -106,19 +107,19 @@ lsregion_to_iovec(const struct lsregion *lsregion, struct iovec *iov,
 		if (lslab->max_id == LSLAB_NOT_USED_ID)
 			break;
 		/* An already seen slab with no new allocations. */
-		if (lslab->max_id <= prev_id)
+		if (lslab->slab_id < prev_id ||
+		    (lslab->slab_id == prev_id &&
+		     lslab->slab_used - lslab_sizeof() == prev_pos)) {
 			continue;
+		}
 		if (cnt >= max_iovcnt)
 			break;
-		if (lslab->max_id >= svp->max_id) {
-			svp->max_id = lslab->max_id;
-			svp->pos = lslab->slab_used - lslab_sizeof();
-		}
-
 		iov->iov_base = (char *)lslab + lslab_sizeof();
 		iov->iov_len = lslab->slab_used - lslab_sizeof();
-
-		if (lslab->min_id <= prev_id) {
+		svp->slab_id = lslab->slab_id;
+		svp->pos = iov->iov_len;
+		max_alloc_id = lslab->max_id;
+		if (lslab->slab_id == prev_id) {
 			/*
 			 * We flushed this slab previously,
 			 * shift the already seen contents.
@@ -132,5 +133,5 @@ lsregion_to_iovec(const struct lsregion *lsregion, struct iovec *iov,
 		cnt++;
 	}
 	*iovcnt = cnt;
-	return;
+	return max_alloc_id;
 }
diff --git a/test/lsregion.c b/test/lsregion.c
index 68ac70e..ab51d27 100644
--- a/test/lsregion.c
+++ b/test/lsregion.c
@@ -535,7 +535,7 @@ test_aligned(void)
 static void
 test_to_iovec(void)
 {
-	plan(17);
+	plan(18);
 	header();
 
 	struct quota quota;
@@ -564,12 +564,12 @@ test_to_iovec(void)
 	is(iov->iov_len, 0);
 
 	iovcnt = 1;
-	lsregion_to_iovec(&allocator, iov, &iovcnt, &svp);
+	int64_t rc_id = lsregion_to_iovec(&allocator, iov, &iovcnt, &svp);
 	/* Check basic return values. */
 	is(iov->iov_len, count * size);
 	is(iovcnt, 1);
 	id += count;
-	is(svp.max_id, id - 1);
+	is(rc_id, id - 1);
 	is(svp.pos, iov->iov_len);
 	/* Check iovec contents. */
 	for (int i = 0; i < count; i++) {
@@ -578,15 +578,17 @@ test_to_iovec(void)
 	test_data(data, count, size);
 	/* Check second flush of the same data */
 	struct lsregion_svp old_svp = svp;
-	lsregion_to_iovec(&allocator, iov, &iovcnt, &svp);
-	is(old_svp.max_id, svp.max_id);
+	int64_t old_id = rc_id;
+	rc_id = lsregion_to_iovec(&allocator, iov, &iovcnt, &svp);
+	is(rc_id, 0);
+	is(old_svp.slab_id, svp.slab_id);
 	is(old_svp.pos, svp.pos);
 	is(iovcnt, 0);
 	/* Check new alloc and flush without gc. */
 	iovcnt = 1;
 	fill_data(data, count, size, id, &allocator);
-	lsregion_to_iovec(&allocator, iov, &iovcnt, &svp);
-	is(svp.max_id, old_svp.max_id + count);
+	rc_id = lsregion_to_iovec(&allocator, iov, &iovcnt, &svp);
+	is(rc_id, old_id + count);
 	is(svp.pos, old_svp.pos + count * size);
 	is(iov->iov_len, count * size);
 	is(iovcnt, 1);
```

</details>